File: src/main/GUIChildModel.java
import java.awt.*;
import java.lang.reflect.InvocationTargetException;
import java.util.concurrent.Callable;

import javax.swing.*;

import org.nlogo.api.*;
import org.nlogo.lite.InterfaceComponent;
import org.nlogo.nvm.HaltException;
import org.nlogo.window.SpeedSliderPanel;
import org.nlogo.workspace.AbstractWorkspace;


public class GUIChildModel extends ChildModel {

    final JFrame frame = new JFrame();
    InterfaceComponent component;
    int levelsSpaceNumber;
    GUIPanel panel;


    public GUIChildModel(World parentWorld, final String path, final int levelsSpaceNumber)
            throws InterruptedException, ExtensionException, HaltException {
        super(parentWorld);
        this.levelsSpaceNumber = levelsSpaceNumber;

        component = runUISafely(new Callable<InterfaceComponent>() {
            public InterfaceComponent call() throws Exception {
                // For some reason, the IC must be added to the frame before the model is opened.
                new JFrame();
                InterfaceComponent component = new InterfaceComponent(frame);
                panel = new GUIPanel(component);
                frame.add(panel);
                frame.setVisible(true);
                component.open(path);
                // get all components, find the speed slider, and hide it.
                Component[] c = component.workspace().viewWidget.controlStrip.getComponents();
                for (Component co : c) {
                    if (co instanceof SpeedSliderPanel) {
                        co.setVisible(false);
                        ((SpeedSliderPanel) co).setValue(0);
                    }
                }
                frame.setTitle(component.workspace().getModelFileName() + " (LevelsSpace model-id: " + String.valueOf(levelsSpaceNumber) + ")");
                frame.pack();
                // Make sure that the model doesn't close if people accidentally click the close button
                frame.setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
                // Adding window listener so that the model calls the method that removes it from
                // the extension if closed.
                frame.addWindowListener(new java.awt.event.WindowAdapter() {
                    @Override
                    public void windowClosing(java.awt.event.WindowEvent windowEvent) {
                        Object[] options = {"Close Model", "Run in Background", "Cancel"};
                        int n = JOptionPane.showOptionDialog(frame,
                                "Close the model, run it in the background, or do nothing?",
                                null, JOptionPane.YES_NO_CANCEL_OPTION,
                                JOptionPane.QUESTION_MESSAGE,
                                null,
                                options,
                                options[2]);
                        switch (n) {
                            case 0:
                                try {
                                    LevelsSpace.closeModel(levelsSpaceNumber);
                                } catch (ExtensionException e) {
                                    throw new RuntimeException(e);
                                } catch (HaltException e) {
                                    //ignore
                                }
                                break;
                            case 1:
                                hide();
                        }
                    }
                });
                return component;
            }
        });
        init();
    }

    public void setSpeed(double d){
        Component[] c = component.workspace().viewWidget.controlStrip.getComponents();
        for (Component co : c){
            if (co instanceof SpeedSliderPanel){
                ((SpeedSliderPanel) co).setValue((int)d);
            }
        }

    }

    @Override
    public AbstractWorkspace workspace() {
        return component.workspace();
    }

    @Override
    JFrame frame() {
        return frame;
    }
}

File: src/main/GUIChildModel.scala
package org.nlogo.ls

import java.awt._
import java.awt.event.{ WindowAdapter, WindowEvent }
import javax.swing.{ JFrame, JMenuBar, WindowConstants }
import java.io.IOException

import org.nlogo.api._
import org.nlogo.app.{ App, ZoomMenu }
import org.nlogo.ls.gui.{ GUIPanel, InterfaceComponent, ZoomableInterfaceComponent }
import org.nlogo.nvm.HaltException
import org.nlogo.swing.{ ModalProgress, NetLogoIcon, Utils }
import org.nlogo.theme.InterfaceColors
import org.nlogo.window.GUIWorkspace

import scala.util.{ Failure, Try }

class GUIChildModel @throws(classOf[InterruptedException]) @throws(classOf[ExtensionException]) @throws(classOf[HaltException]) @throws(classOf[IOException])
(ls: LevelSpace, parentWorkspace: Workspace, path: String, modelID: Int)
  extends ChildModel(parentWorkspace, modelID) {

  val (component, panel, frame) = UnlockAndBlock.onEDT(parentWorkspace.world) {
    val f = new JFrame with NetLogoIcon with ModalProgress
    val component: InterfaceComponent = new ZoomableInterfaceComponent(f)
    val panel = new GUIPanel(component.workspace, component)
    (component, panel, Some(f))
  }

  var menuBar = new SyncedMenuBar()

  UnlockAndBlock.onEDT(parentWorkspace.world) (Try {
    val f = frame.get
    f.add(panel)
    val currentlyFocused: Window =
      Option(KeyboardFocusManager.getCurrentKeyboardFocusManager.getActiveWindow).getOrElse(App.app.frame)
    f.setLocationRelativeTo(currentlyFocused)
    f.setLocationByPlatform(true)
    f.setVisible(true)
    currentlyFocused.toFront()
    openModelWithoutGenerator(component.open, path)
    panel.packSplitPane()
    f.pack()
    panel.resetCommandCenter()
    f.setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE)
    f.addWindowListener(new GUIWindowAdapter)
    f.setJMenuBar(menuBar)
    (component, panel)
  }) match {
    case Failure(_) => kill()
    case _ =>
  }
  updateFrameTitle()
  syncTheme()

  class GUIWindowAdapter extends WindowAdapter {
    override def windowClosing(windowEvent: WindowEvent): Unit = hide()
  }

  def setSpeed(d: Double): Unit = {
    // Note the panel will update and store the speed even if the frame is not visible. The thing is, we don't want
    // to slow the model down if it's not visible
    workspace.updateManager.speed = d
    // Wakes up the workspace if the speed slider is super low.
    // Makes it so there's not a long pause after increasing
    // the speed slider from a low position. BCH 6/18/2016
    workspace.updateManager.nudgeSleeper()
    onEDT {
      panel.speedSlider.setValue((d * 2).intValue)
    }
  }

  override def hide(): Unit = {
    workspace.updateManager.speed = 50
    workspace.updateManager.nudgeSleeper()
    super.hide()
  }

  override def show(): Unit = {
    workspace.updateManager.speed = panel.speedSlider.getValue / 2.0
    syncTheme()
    super.show()
  }

  def workspace: GUIWorkspace = component.workspace

  def syncTheme(): Unit = {
    menuBar.syncTheme()
    panel.syncTheme()

    frame.foreach(_.repaint())
  }
}

class SyncedMenuBar extends JMenuBar {
  // val zoomMenuClass = Class.forName("org.nlogo.app.ZoomMenu")
  // add(zoomMenuClass.getDeclaredConstructor().newInstance().asInstanceOf[Menu])

  private val zoomMenu = new ZoomMenu

  add(zoomMenu)

  override def paintComponent(g: Graphics): Unit = {
    val g2d = Utils.initGraphics2D(g)

    g2d.setColor(InterfaceColors.menuBackground())
    g2d.fillRect(0, 0, getWidth, getHeight)
  }

  override def paintBorder(g: Graphics): Unit = {
    val g2d = Utils.initGraphics2D(g)

    g2d.setColor(InterfaceColors.menuBarBorder())
    g2d.drawLine(0, getHeight - 1, getWidth, getHeight - 1)
  }

  def syncTheme(): Unit = {
    zoomMenu.syncTheme()
  }
}
